bitcoin/src/blockdata/transaction.rs

```diff
@@ -225,6 +225,28 @@ impl TxIn {
     pub fn enables_lock_time(&self) -> bool {
         self.sequence != Sequence::MAX
     }
+
+    /// The weight of the TxIn when it's included in a legacy transaction (i.e., a transaction
+    /// having only legacy inputs).
+    ///
+    /// The witness weight is ignored here even when the witness is non-empty.
+    /// If you want the witness to be taken into account, use `TxIn::segwit_weight` instead.
+    pub fn legacy_weight(&self) -> usize {
+        let script_sig_size = self.script_sig.len();
+        // Size in vbytes:
+        // previous_output (36) + script_sig varint len + script_sig push + sequence (4)
+        // We then multiply by 4 to convert to WU
+        (36 + VarInt(script_sig_size as u64).len() + script_sig_size + 4) * 4
+    }
+
+    /// The weight of the TxIn when it's included in a segwit transaction (i.e., a transcation
+    /// having at least one segwit input).
+    ///
+    /// This always takes into account the witness, even when empty, in which
+    /// case 1WU for the witness length varint (`00`) is included.
+    pub fn segwit_weight(&self) -> usize {
```

Comment:
If this is not called `min_segwit_weight` (why?) it should have a well-visible warning that it doesn't account for varint changing so naive sum wouldn't work.

Also I recently found out that Core has `FeeRate` type but we don't which feels ridiculous. I obviously don't ask you to add it just a note.

Reply:
Right! I just added a warning to the docs, let me know how you feel about it :)

```diff
+    /// Keep in mind that when adding a TxIn to a transaction, the total weight of the transaction
+    /// might increase more than `TxIn::segwit_weight`. This happens when:
+    /// - the new input added causes the input length `VarInt` to increase its encoding length
+    /// - the new input is the first segwit input added - this will add an additional 2WU to the
+    /// transaction weight to take into account the segwit marker
```